File: frontend/onied/src/components/blocks/video/providers/youtubeVideoProvider.ts
import VideoProvider from "./videoProvider";

/**
 * Youtube Video Provider.
 *
 * @class YoutubeVideoProvider
 * @extends {VideoProvider}
 */
export default class YoutubeVideoProvider extends VideoProvider {
  rawVideo = false;
  regex =
    /^((?:https?:)?\/\/)?((?:www|m).)?((?:youtube(-nocookie)?.com|youtu.be))(\/(?:[\w-]+\?v=|embed\/|live\/|v\/)?)(?<videoId>[\w]+)(\S+)?$/;

  getLink(href: string) {
    const matches = href.match(this.regex);

    return Promise.resolve(
      `https://www.youtube.com/embed/${matches?.groups?.videoId}`
    );
  }
}

File: mobile/onied/lib/pages/course/components/blocks/videoBlock/video/providers/youtube_video_provider.dart
import 'video_provider.dart';

class YouTubeVideoProvider extends VideoProvider {
  @override
  RegExp get regex => RegExp(
    r'^((?:https?:)?\/\/)?((?:www|m).)?((?:youtube(-nocookie)?.com|youtu.be))(\/(?:[\w-]+\?v=|embed\/|live\/|v\/)?)(?<videoId>[\w]+)(\S+)?$',
  );

  @override
  String getLink(String href) {
    final matches = regex.firstMatch(href);
    if (matches == null || matches.namedGroup('videoId') == null) {
      throw Exception('Невалидный URL YouTube');
    }

    final videoId = matches.namedGroup('videoId')!;
    return 'https://www.youtube.com/embed/$videoId';
  }
}
